onnxruntime/python/onnxruntime_pybind_ortvalue.cc

```diff
@@ -213,12 +228,37 @@ void addOrtValueMethods(pybind11::module& m) {
 #ifdef ENABLE_TRAINING
       .def("to_dlpack", [](OrtValue* ort_value) -> py::object {
         return py::reinterpret_steal<py::object>(ToDlpack(*ort_value));
-      })
-      .def_static("from_dlpack", [](py::object data, bool is_bool_tensor = false) {
+      }, "Returns a DLPack representing the tensor. This method does not copy the pointer shape, "
+         "instead, it copies the pointer value. The OrtValue must be persist until the dlpack structure "
+         "is consumed.")
+      .def_static("from_dlpack", [](py::object data, bool is_bool_tensor) {
         return FromDlpack(data.ptr(), is_bool_tensor);
-      })
+      }, py::arg("data"), py::arg("is_bool_tensor")=false,
+        "Converts a tensor from a external library into an OrtValue by means of the __dlpack__ protocol.")
+      .def("__dlpack__", [](OrtValue* ort_value, py::object /* stream */) -> py::object {
+        return py::reinterpret_steal<py::object>(ToDlpack(*ort_value));
+       }, py::arg("stream")=py::none(),
+       "Returns a DLPack representing the tensor (part of __dlpack__ protocol). "
+       "This method does not copy the pointer shape, instead, it copies the pointer value. "
+       "The OrtValue must persist until the dlpack structure is consumed.")
+      .def("__dlpack_device__", [](const OrtValue* ort_value) -> py::tuple {
+        ORT_ENFORCE(ort_value->IsTensor(), "Only tensor type OrtValues are supported");
+        const onnxruntime::Tensor& tensor = ort_value->Get<Tensor>();
+        DLDevice device = onnxruntime::dlpack::GetDlpackDevice(*ort_value, tensor.Location().device.Id());
+        return py::make_tuple(static_cast<int>(device.device_type), device.device_id);
+       }, "Returns a tuple of integers, (device, device index) (part of __dlpack__ protocol).")
 #endif
       ;
+
+#ifdef ENABLE_TRAINING
+  m.def("may_dlpack_bool_tensor", [](py::capsule cap) -> bool {
```

Comment:
Why do we need this function? And it's confusing from the name and the doc here. The condition here means it's possible to be UINT8 or BOOL, so the "may" in the name seems reasonable, but the doc "Tells if a DLPack structure is a boolean tensor." is confusing. This function is used only once (ignoring the test file) at "_utils.py" below: `if C.may_dlpack_bool_tensor(dlp) and ortvalues[i].proto_type() == 9:`, but it's not possible that proto_type is 9 but may_dlpack_bool_tensor returns false. We should not use IF to hide problem, if we really want to do some check, we should assert.